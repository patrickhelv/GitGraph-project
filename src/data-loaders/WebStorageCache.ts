import { AxiosResponse } from 'axios'

export class WebStorageCache {
  public setCahce<T>(key: string, response: AxiosResponse<T>) {
    window.sessionStorage.setItem(key, JSON.stringify(response.data))
  }

  public getCache<T>(key: string): AxiosResponse<T> | undefined {
    const cache = window.sessionStorage.getItem(key)
    if (!cache) {
      return undefined
    }
    const sessionCache = JSON.parse(cache) as T

    return this.mapAxiosResponse(200, 'success', sessionCache)
  }

  private mapAxiosResponse<T>(
    status: number,
    statusText: string,
    data: T | undefined
  ): AxiosResponse<T> {
    return {
      headers: {},
      config: {},
      status,
      statusText,
      data,
    } as AxiosResponse<T>
  }
}
