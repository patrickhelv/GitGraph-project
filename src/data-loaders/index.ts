export * from './APILoader'
export * from './WebStorageCache'
