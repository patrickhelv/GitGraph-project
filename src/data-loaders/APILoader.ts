import {
  IGitLabIssue,
  IGitLabUser,
  IGitLabCommit,
  IGitLabContributor,
} from '../content/types'
import axios from '../http-common'
import { AxiosError, AxiosResponse } from 'axios'
import { User, Issue, Commit, Contributor } from '../models'
import { arrayOrUndefined } from '../utils/arrayOrUndefined'
import { WebStorageCache } from './WebStorageCache'

export class APILoader {
  readonly cahce: WebStorageCache

  constructor() {
    this.cahce = new WebStorageCache()
  }

  private async load<T>(parameter: string): Promise<AxiosResponse<T>> {
    return axios.get<T>(parameter)
  }

  private getCache<T>(key: string): AxiosResponse<T> | undefined {
    return this.cahce.getCache(key)
  }

  private async getFeed<T>(key: string): Promise<AxiosResponse<T>> {
    const sessionChace = this.getCache<T>(key)

    if (!sessionChace) {
      const response = await this.load<T>(key)
      this.cahce.setCahce(key, response)
      return response
    }
    return sessionChace
  }

  public async getIssues(): Promise<Array<Issue> | undefined | string> {
    try {
      const response = await this.getFeed<IGitLabIssue[]>('issues')

      return arrayOrUndefined(
        response.data.map((issue: IGitLabIssue) => {
          return new Issue({
            id: issue.id,
            title: issue.title,
            description: issue.description,
            assignees: this.getIssueAssignee(issue),
            state: issue.state,
          })
        })
      )
    } catch (e) {
      const error = e as AxiosError
      if (!error.response) {
        return undefined
      }
      return `Status: ${error.response.status}`
    }
  }

  private getIssueAssignee(issue: IGitLabIssue): User[] | undefined {
    if (!issue.assignees) {
      return undefined
    }
    return arrayOrUndefined(
      issue.assignees.map((user: IGitLabUser) => {
        return new User({
          id: user.id,
          name: user.name,
          avatarUrl: user.avatar_url,
        })
      })
    )
  }

  public async getCommits(): Promise<Array<Commit> | undefined | string> {
    try {
      const response = await this.getFeed<IGitLabCommit[]>(
        'repository/commits?per_page=100'
      )

      return arrayOrUndefined(
        response.data.map((commit: IGitLabCommit) => {
          return new Commit({
            id: commit.short_id,
            title: commit.title,
            committerName: commit.committer_name,
            createdAt: commit.created_at,
            message: commit.message,
            webUrl: commit.web_url,
          })
        })
      )
    } catch (e) {
      const error = e as AxiosError
      if (!error.response) {
        return undefined
      }
      return `Status: ${error.response.status}`
    }
  }

  public async getUsers(): Promise<Array<User> | undefined | string> {
    try {
      const response = await this.getFeed<IGitLabUser[]>('users')

      return arrayOrUndefined(
        response.data.map((user: IGitLabUser) => {
          return new User({
            id: user.id,
            name: user.name,
            avatarUrl: user.avatar_url,
          })
        })
      )
    } catch (e) {
      const error = e as AxiosError
      if (!error.response) {
        return undefined
      }
      return `Status: ${error.response.status}`
    }
  }

  public async getUserIssues(user: User): Promise<Array<Issue> | undefined> {
    // TODO add cache/state
    const response = await this.getIssues()
    if (!response) {
      return undefined
    }
    if (typeof response === 'string') {
      return undefined
    }

    arrayOrUndefined(
      response.filter((issue) => {
        const assigneeIds = issue.getAssigneeIds()
        if (assigneeIds) {
          return assigneeIds.includes(user.getId())
        }
        return false
      })
    )
  }

  public async geContributors(): Promise<
    Array<Contributor> | undefined | string
  > {
    try {
      const response = await this.getFeed<IGitLabContributor[]>(
        'repository/contributors'
      )

      return arrayOrUndefined(
        response.data.map((contributor: IGitLabContributor) => {
          return new Contributor({
            commits: contributor.commits,
            email: contributor.email,
            name: contributor.name,
          })
        })
      )
    } catch (e) {
      const error = e as AxiosError
      if (!error.response) {
        return undefined
      }
      return `Status: ${error.response.status}`
    }
  }

  public async getPrecentageIssuesCompleted(): Promise<number | undefined> {
    const issues = await this.getIssues()

    if (!issues || typeof issues === 'string') {
      return undefined
    }

    return (
      issues.filter((issue) => issue.getState() === 'closed').length /
      issues.length
    )
  }
}
