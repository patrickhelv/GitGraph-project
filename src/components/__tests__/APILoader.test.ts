import { APILoader } from '../../data-loaders'

const loader = new APILoader()
const TIMEOUT = 10000

async function testClosedIssues() {
  const response = await loader.getIssues()
  if (typeof response === 'string') {
    throw Error(response)
  }

  if (response) {
    const closedIssues = response.filter(
      (issue) => issue.getState() === 'closed'
    )
    expect(closedIssues).toMatchSnapshot()
  }
}

describe('APILoader test', () => {
  it('Test closed issues', () => testClosedIssues(), TIMEOUT)
})
