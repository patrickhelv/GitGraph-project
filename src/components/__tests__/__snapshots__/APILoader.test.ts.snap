// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APILoader test Test closed issues 1`] = `
Array [
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": "Use Reat Context API, for example to interact with toggleButton",
      "id": 54568,
      "state": "closed",
      "title": "Use React Context API",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/2350/avatar.png",
            "id": 2350,
            "name": "Marius Arhaug",
          },
        },
      ],
      "description": "right now info box is fugyl, fix it",
      "id": 54552,
      "state": "closed",
      "title": "Refactor info box",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/1138/avatar.png",
            "id": 1138,
            "name": "Patrick Øivind Helvik Legendre",
          },
        },
      ],
      "description": null,
      "id": 54461,
      "state": "closed",
      "title": "FEAT: Button Last 5 days",
    },
  },
  Issue {
    "props": Object {
      "assignees": undefined,
      "description": null,
      "id": 54221,
      "state": "closed",
      "title": "Fix footer",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/2350/avatar.png",
            "id": 2350,
            "name": "Marius Arhaug",
          },
        },
      ],
      "description": null,
      "id": 54220,
      "state": "closed",
      "title": "Fix infobox",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/2350/avatar.png",
            "id": 2350,
            "name": "Marius Arhaug",
          },
        },
      ],
      "description": "Make API loader a single structure, make parent component, react class components with state and make individual components functional components without much extra features.",
      "id": 54134,
      "state": "closed",
      "title": "CHANGE: refactor code base strucutre",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": null,
      "id": 54126,
      "state": "closed",
      "title": "FEAT: WebStorage",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/2350/avatar.png",
            "id": 2350,
            "name": "Marius Arhaug",
          },
        },
      ],
      "description": null,
      "id": 54125,
      "state": "closed",
      "title": "Write comments for documentation",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": "Testing website on 3 different devices computer, mobile, monitor. Document the testing afterwards.",
      "id": 54124,
      "state": "closed",
      "title": "TEST: frontend and responsive design",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": null,
      "id": 54123,
      "state": "closed",
      "title": "TEST: create a test for component behaviour",
    },
  },
  Issue {
    "props": Object {
      "assignees": undefined,
      "description": "APIManager is run and rerenders the DOM 4 times on each refresh?
Could this be due to to the way useEffect is used with the API functions and useState?",
      "id": 54034,
      "state": "closed",
      "title": "API requesting multiple times",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/da38006dcacf8156bce26d14186fe363?s=80&d=identicon",
            "id": 3157,
            "name": "Ole Alexander Høyby",
          },
        },
      ],
      "description": null,
      "id": 54026,
      "state": "closed",
      "title": "CHANGE: Revise documentation",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": null,
      "id": 53800,
      "state": "closed",
      "title": "FEAT: Create last 5 issues",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/da38006dcacf8156bce26d14186fe363?s=80&d=identicon",
            "id": 3157,
            "name": "Ole Alexander Høyby",
          },
        },
      ],
      "description": null,
      "id": 53799,
      "state": "closed",
      "title": "FEAT: Create filter for main graph commits",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": "",
      "id": 53734,
      "state": "closed",
      "title": "FEAT: Save axios requests in sessionstorage",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/2350/avatar.png",
            "id": 2350,
            "name": "Marius Arhaug",
          },
        },
      ],
      "description": null,
      "id": 53731,
      "state": "closed",
      "title": "FEAT: % issues resolved",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/1138/avatar.png",
            "id": 1138,
            "name": "Patrick Øivind Helvik Legendre",
          },
        },
      ],
      "description": null,
      "id": 53696,
      "state": "closed",
      "title": "FEAT: create infobox",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://secure.gravatar.com/avatar/e5e39c38cd6b4b353a152b804f90566b?s=80&d=identicon",
            "id": 3197,
            "name": "Stefan Djordje Tomic",
          },
        },
      ],
      "description": "* [x] Write snapshot-tests
* [ ] Write API tests using Axios",
      "id": 52904,
      "state": "closed",
      "title": "FEAT: Expand testing",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/2350/avatar.png",
            "id": 2350,
            "name": "Marius Arhaug",
          },
        },
      ],
      "description": "at it2810-XX.idi.ntnu.no

Forelesning 3 om dette prosjektet sier vi skal sette opp websever?",
      "id": 52783,
      "state": "closed",
      "title": "FEAT: Set up webserver",
    },
  },
  Issue {
    "props": Object {
      "assignees": Array [
        User {
          "props": Object {
            "avatarUrl": "https://gitlab.stud.idi.ntnu.no/uploads/-/system/user/avatar/1138/avatar.png",
            "id": 1138,
            "name": "Patrick Øivind Helvik Legendre",
          },
        },
      ],
      "description": null,
      "id": 52485,
      "state": "closed",
      "title": "FEAT: create charts",
    },
  },
]
`;
